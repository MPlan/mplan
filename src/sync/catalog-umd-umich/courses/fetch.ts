import axios from 'axios';

export async function fetchCourseList(level: 'undergraduate' | 'graduate', subjectCode: string) {
  const response = await axios({
    url: `http://catalog.umd.umich.edu/${level}/coursesaz/${subjectCode.toLowerCase().trim()}`,
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });
  const html = response.data as string | undefined;
  if (!html) throw new Error(`could not get course list for "${level} ${subjectCode}"`);
  return html;
}
