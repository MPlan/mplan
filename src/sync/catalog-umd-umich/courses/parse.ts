// import { Course } from 'sync/catalog-umd-umich/models';
import { JSDOM } from 'jsdom';
import { parseCourseBlockTitle } from './parse-course-block-title';
import { parseCourseBlockExtra } from './parse-course-block-extra';
import { Prerequisite, Course } from '../models';

export function parseCourses(courseListHtml: string): Course[] {
  const { window } = new JSDOM(courseListHtml);
  const { document } = window;
  const courseBlocks = Array.from(document.querySelectorAll('.courseblock'));
  return courseBlocks.map(courseBlock => {
    // subject code, course number, name, and credit hours
    const courseBlockTitleEl = courseBlock.querySelector('.courseblocktitle');
    if (!courseBlockTitleEl) throw new Error('could not find .courseblocktitle');
    const courseBlockTitle = courseBlockTitleEl.textContent;
    if (!courseBlockTitle) throw new Error('could not get text content of .courseblocktitle');
    const { courseNumber, creditHours, name, subjectCode } = parseCourseBlockTitle(
      courseBlockTitle,
    );

    // description
    const descriptionEl = courseBlock.querySelector('.courseblockdesc');
    if (!descriptionEl) {
      throw new Error('could not get description because could not find .courseblockdesc');
    }
    const _description = descriptionEl.textContent;
    if (!_description) throw new Error('description block text content was falsy');
    const description = _description.trim();

    const extras = Array.from(courseBlock.querySelectorAll('.courseblockextra'))
      .map(block => {
        try {
          return parseCourseBlockExtra(block);
        } catch (e) {
          console.warn(`Couldn't parse ${subjectCode} ${courseNumber}`, e);
          return undefined;
        }
      })
      .filter(x => !!x)
      .reduce(
        (extras, next) => {
          return { ...extras, ...next };
        },
        {} as any,
      ) as { prerequisites?: Prerequisite; corequisites?: Prerequisite; restrictions?: string[] };

    return {
      subjectCode,
      courseNumber,
      name,
      creditHours,
      description,
      ...extras,
    };
  });
}
