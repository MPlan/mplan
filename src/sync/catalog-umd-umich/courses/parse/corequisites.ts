export function parseCorequisites(courseBlockExtra: Element) {
  
}