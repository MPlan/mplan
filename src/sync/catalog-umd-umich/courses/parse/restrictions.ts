export function parseRestrictionsBlock(restrictionsBlock: Element) {
  return Array.from(restrictionsBlock.childNodes)
    .map(x => x.textContent!)
    .filter(x => !!x)
    .map(x => x.trim())
    .filter(x => !!x)
    .filter(x => !/restriction/i.test(x));
}
