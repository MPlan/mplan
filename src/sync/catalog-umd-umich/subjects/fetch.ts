import axios from 'axios';

export async function fetchUndergraduateSubjectsHtml() {
  const response = await axios({
    method: 'GET',
    url: 'http://catalog.umd.umich.edu/undergraduate/coursesaz/',
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) {
    throw new Error('Subjects response was undefined or empty.');
  }
  return html;
}
