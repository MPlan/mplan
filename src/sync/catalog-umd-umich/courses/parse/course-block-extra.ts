import { parsePrerequisites } from './prerequisites';
import { parseRestrictionsBlock } from './restrictions';

export function parseCourseBlockExtra(courseBlockExtra: Element) {
  const textContent = courseBlockExtra.textContent;
  if (!textContent) throw new Error('could not get textContent from .courseblockextra');
  if (/restriction/i.test(textContent)) {
    return { restrictions: parseRestrictionsBlock(courseBlockExtra) };
  }
  if (/prerequisite/i.test(textContent)) {
    return { prerequisites: parsePrerequisites(courseBlockExtra) };
  }
  if (/corequisite/i.test(textContent)) {
    return { corequisites: parsePrerequisites(courseBlockExtra) };
  }
  return undefined;
}
