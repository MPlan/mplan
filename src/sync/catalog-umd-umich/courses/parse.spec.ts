import { getOrWriteHtml } from 'testing-utilities/example-html';
import { parseCourses } from './parse';
import { fetchCourseList } from './fetch';

describe('courses', () => {
  describe('parse', () => {
    it('parses the CIS course page', async () => {
      const html = await getOrWriteHtml('cis-courses', __dirname, () =>
        fetchCourseList('undergraduate', 'cis'),
      );

      const courses = parseCourses(html);
      expect(courses).toMatchSnapshot();
    });
  });
});
