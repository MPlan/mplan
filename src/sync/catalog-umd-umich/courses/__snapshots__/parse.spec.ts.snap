// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`courses parse parses the CIS course page 1`] = `
Array [
  Object {
    "courseNumber": "112",
    "creditHours": 3,
    "description": "This is a microcomputer literacy course with primary emphasis on the application tools of the word processor, spreadsheets, and database.  Additional topics of computer terms, systems, and use in society are included.  The course is intended for undergraduates in the College of Arts, Sciences, and Letters.  No previous experience with computers is expected. (YR).",
    "name": "Computer Literacy/Info Mgmt",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "125",
    "creditHours": 3,
    "description": "A survey of computer science topics, including history of computing, office productivity software, the internet, HTML, JavaScript, web design, algorithms, assemblers and compilers, gates and logic design, models of computation, artificial intelligence and expert systems, computing ethics, privacy issues, intellectual property. No credit for CIS majors. (F,W,S).",
    "name": "Survey of Computer Science",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "150",
    "creditHours": 4,
    "description": "This course provides a foundation for further studies in computer and information science and emphasizes a structured approach to problem solving and algorithm development.  Topics include principles of program design, coding, debugging, testing, and documentation.  Students are introduced to the Unified Modeling Language for requirements analysis using use-cases and activity diagrams, an object oriented programming language, and the fundamentals of computer hardware, system software, and components.  The course will consist of three lecture hours and one two-hour laboratory.",
    "name": "Computer Science I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "1501",
    "creditHours": 4,
    "description": "This course provides a foundation for further studies in computer and information science and emphasizes a structured approach to problem solving and algorithm development using a high-level language more suited to data science applications. Topics include principles of program design, coding, debugging, testing, and documentation. Students are introduced to the Unified Modeling Language for requirements analysis using use-cases and activity diagrams, an object-oriented programming language for data science applications, and the fundamentals of computer hardware, system software, and components. The course will consist of three lecture hours and one two-hour laboratory. The labs will cover various data science applications. (F,W,S)",
    "name": "CS I for Data Scientists",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "167",
    "creditHours": 3,
    "description": "This course wil lcover installation, configuration, and troubleshooting Microsoft Proxy Server, methods of controlling internet access, ehancement to performance monitor, and creation of a plan for integrating Proxy Server in an existing environment. (F,W)",
    "name": "Proxy Server",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "200",
    "creditHours": 4,
    "description": "This course presents techniques for the design, writing, testing, and debugging of medium-sized programs, and an introduction to data structures (stacks, queues, linked lists) using an object-orientated programming language.  Topics covered include pointers, templates, and inheritance.  The principles of UML modeling are continued.  This course will consist of three lecture hours and one two-hour laboratory.",
    "name": "Computer Science II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "2001",
    "creditHours": 4,
    "description": "This course presents techniques for the design, writing, testing, and debugging of medium-sized programs, and an introduction to data structures (stacks, queues, linked lists) using an object-oriented programming language for data science applications. Topics covered include pointers, templates, and inheritance. The principles of UML modeling are continued. This course will consist of three lecture hours and one two-hour laboratory. The labs will cover various data science applications. (F, W, S)",
    "name": "CS II for Data Scientists",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "205",
    "creditHours": 3,
    "description": "Full Course Title: Computer Programming for Engineers- Intermediate topics in computer programming: arrays, files, structured data types, pointers, functions. Overview of digital computer hardware and system software components: machine architecture, operating systems, computer networks, data security, and performance evaluation. No credit for CIS majors.",
    "name": "Comp Programming for Engineers",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "275",
    "creditHours": 4,
    "description": "This course introduces students to various topics in discrete mathematics, such as set theory, mathematical logic, trees, and graph theory. Applications to relational databases, modeling reactive systems and program verification are also discussed. (F,W,S)",
    "name": "Discrete Structures I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "280",
    "creditHours": 3,
    "description": "This course provides a practical introduction to a broad range of computer security topics.  Covered topics include: practical computer security principles to help empower students to secure their own connections to cyberspace; firewalls, malware, and intrusion detection; cryptography basics and its applications; mobile devices and related security issues; network technologies and their vulnerabilities.",
    "name": "Prac Aspects of Computer Sec",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "285",
    "creditHours": 3,
    "description": "This course will cover various CASE tools, such as UML modeling and code generation tools, configuration management tools, defect management tools, an integrated development environment for coding and debugging, unit and testing tools, and build tools. Students will learn these tools in a laboratory environment. This course will be comprised of one lecture hour and one two-hour laboratory. (F,W)",
    "name": "Software Engineering Tools",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "290",
    "creditHours": 2,
    "description": "One significant programming language is covered in depth. The particular language changes from term to term. The language chosen might be Ada, C, MODULA 2, USP, PROLOG, or SMALLTALK.",
    "name": "Topic in Programming Languages",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "290A",
    "creditHours": 2,
    "description": "TOPIC TITLE: Advanced Visual C++ Programming. Advanced treatment of the Visual C++ Programming.",
    "name": "Topic in Programming Languages",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "294",
    "creditHours": 3,
    "description": "An introduction to create professional-looking applications using the graphical user interface of Windows. Students learn how to create graphical objects and controls, write event driven code that responds to clicking on buttons, work with multiple forms and executable files. (F,S).",
    "name": "Programming with Visual Basic",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "296",
    "creditHours": 3,
    "description": "Course covers Java Programming language, focusing on GUI development, distributed computing and network applications.",
    "name": "Java Programming",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "297",
    "creditHours": 3,
    "description": "This course provides an introduction to the C# programming language and the .NET Framework for the development of Windows game applications. Some discussion of DirectX programming and Xbox game development is also included. (W)",
    "name": "Intro to C Sharp",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "299",
    "creditHours": 1,
    "description": "Student works with an industrial sponsor in the area of CIS. Approval of Internship Coordinator required. (F,W,S).",
    "name": "Internship",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "306",
    "creditHours": 4,
    "description": "This course introduces students to further topics in discrete mathematics, including theory of computation, more complexity theory, coding theory, and game theory.",
    "name": "Discrete Structures II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "310",
    "creditHours": 4,
    "description": "The architecture of computer systems and associated software. Topics include digital logic circuits, computer interfacing, interrupt systems, input/output systems, memory systems, assemblers and assembly language programming, and computer networks. (F,W,S).",
    "name": "Computer Org and Assembly Lang",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "3200",
    "creditHours": 4,
    "description": "This course provides an overview of what Big Data is and explores its characteristics. It introduces the fundamental technologies, platforms, and methods that enable Big Data analysis, and covers how to acquire, store, and analyze very large amounts of information to complete Big Data analysis tasks. Topics include MapReduce, similarity search, mining real-time data streams, link analysis, clustering, recommender systems, social network graph mining, and large scale data mining tasks. (W)",
    "name": "Data Science II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "350",
    "creditHours": 4,
    "description": "This course focuses on data design and algorithm design. Data design topics include object-oriented discussions of hashing, advanced tree structures, graphs, and sets.  Algorithm design topics include the greedy, divide-and-conquer, dynamic programming, backtracking and branch-and-bound techniques. A significant discussion of algorithm complexity theory, including time and space trade-offs and elementary computability theory, is included. (F,W,S)",
    "name": "Data Struc and Algorithm Anlys",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "3501",
    "creditHours": 4,
    "description": "This course focuses on data design and algorithm design for software engineers. Data design topics include object-oriented discussions of hashing, advanced tree structures, graphs and sets. Algorithm design topics include the greedy, divide-and-conquer, dynamic programming, backtracking and branch-and-bound techniques. A significant discussion of algorithm complexity theory, including time and space trade-offs and elementary computability theory, is included. (F,W,S)",
    "name": "Data Struc & Alg Anlys for SE",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "375",
    "creditHours": 4,
    "description": "This course presents an in-depth treatment of the following software engineering topics: software engineering paradigms, requirements, specification, functional design, object-oriented design, user interface design, software verification and validation, and the maintenance and management of software engineering artifacts, as well as an introductory discussion of software reliability. Various phases of the software engineering process will be modeled using UML. (F,W)",
    "name": "Software Engineering I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "376",
    "creditHours": 4,
    "description": "This course continues the formal development of the software engineering material begun in CIS 375. Topics covered include personal software process, team software process, formal methods, security, software architecture, software quality assurance, software fault tolerance, the evaluation of the effectiveness of human computer interaction and software reliability. (W,S)",
    "name": "Software Engineering II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "381",
    "creditHours": 4,
    "description": "The course introduces students in engineering and computer science to fundamentals of robotics technology, programming and their applications in industrial environment. The emphasis will be on robotics anatomy and configurations, robotocs kinematics, end effectors, use of sensors in robotics, robotics programming, design of robot workcell, robotics applications to production problems, cost justifications and robotics safety, rather than on the extensive theory of robotics. Three-hour lecture and three-hour laboratory per week.",
    "name": "Industrial Robots",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "387",
    "creditHours": 4,
    "description": "This course takes a detailed, hands-on approach to study the procedures and techniques used to identify, extract, validate, document and preserve electronic evidence. Students completing this course will be familiar with the core computer science theory and practical skills necessary to perform basic computer forensic investigations, understand the role of technology in investigating computer-based crime, and be prepared to deal with investigative bodies at a basic level.",
    "name": "Digital Forensics I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "390",
    "creditHours": Array [
      1,
      3,
    ],
    "description": "A course designed to offer selected topics in an area of computer science. The specific topics will be announced (together with special prerequisites) each time offered. Students must elect different topics to take both CIS 390 and CIS 391. (OC).",
    "name": "Topics in Computer Science",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "391",
    "creditHours": Array [
      1,
      3,
    ],
    "description": "A course designed to offer selected topics in an area of computer science. The specific topics will be announced (together with special prerequisites) each time offered. Students must elect different topics to take both CIS 390 and CIS 391. (OC).",
    "name": "Topics in Computer Science II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "399",
    "creditHours": 1,
    "description": "Student works with industrial sponsor in the area of CIS. Permission of Internship Coordinator required. (F,W,S).",
    "name": "Internship",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "400",
    "creditHours": 4,
    "description": "Systematic study of programming languages with regard to their implementation, structures, and use. Languages are compared with regard to their various data types, data structures, operations, control structures, programming environments, and ease of use in solving various programming problems. (F,W).",
    "name": "Programming Languages",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "405",
    "creditHours": 3,
    "description": "This course investigates how to design efficient algorithms. Topics include asymptotic analysis, amortized analysis, divide-and-conquer, dynamic programming, greedy algorithms, branch and bound, backtracking, lower bounds, NP-completeness and approximation algorithms.",
    "name": "Algorithm Analysis & Design",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "421",
    "creditHours": 4,
    "description": "An introduction to database systems, concepts, and techniques. Topics covered include: database environments, ER modeling, relational data model, object-oriented databases, database design theory and methodology, database languages, query processing and optimization, concurrency control, database recovery, and database security.",
    "name": "Database Mgmt Systems",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "422",
    "creditHours": 4,
    "description": "An introduction to database systems, concepts, and techniques for big data. The course discusses classical relational technologies, and then covers the more current approaches to managing massive amounts of data for analytics purposes. Topics include database environments, database design, the relational data model, normalization, SQL, query processing, parallel databases and query processing, in-database analytics, data warehousing, key-value and column stores, NoSQL and NewSQL approaches for managing massive data. (F)",
    "name": "Massive Data Management",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "423",
    "creditHours": 3,
    "description": "The application of artificial intelligence to building decision support and expert systems for management and other applications. Topics include fundamentals of artificial intelligence, knowledge representation and knowledge processing, tools for building expert systems (logic programming, expert shells), decision support system design (modeling and simulation), expert system design (knowledge engineering, learning). (F).",
    "name": "Dec Support and Exp Systems",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "425",
    "creditHours": 4,
    "description": "This course is an introduction to the principles of information systems analysis and design and their role in business organizations. Topics include Systems Development Life Cycle (SDLC), using CASE (Computer Aided Software Engineering) tools for systems design and analysis, prototyping, Rapid Application Development (RAD), extreme programming, quality assurance through software engineering, and object-oriented systems design and analysis using UML (Unified Modeling Language). Participation in a major design project is a requirement for this course.",
    "name": "Information Systems",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4261",
    "creditHours": 4,
    "description": "An introduction to the principles of information systems analysis and design and their role in business organizations. Topics include information systems strategy and planning, ethical issues in information systems, system modeling, clean-room system engineering, domain ontologies, UML, Enterprise Unified Process, e-business, and supply-chain management, deployment and support. Participation in a major design project is a requirement for this course. (F).",
    "name": "Inf Sys Analysis & Design I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4262",
    "creditHours": 4,
    "description": "This course is a continuation of CIS 4261 and provides students with breadth and depth in the information systems area. Topics include web-based information systems, e-commerce, computer-supported collaborative work, workflow systems, data mining, and data warehousing. Participation in a major design project is a requirement of this course. (W).",
    "name": "Inf Sys Analysis & Design II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "427",
    "creditHours": 4,
    "description": "Study of the management aspects of computing networks and distributed systems. Topics include network architectures (ISO/OSI, TCP/IP, ATM), communication hardware (transmission media, network adaptors, switches), encoding, framing, error detection and correction, reliable transmission, data link control and LAN technology, internetworking, routing/congestion control, network design/management.",
    "name": "Comp Networks and Dis Process",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "428",
    "creditHours": 3,
    "description": "The course requires students to setup and manage their own computer network in the lab. Topics include: overview of file servers, LAN configurations and protocols, server hardware (CPU, hard drives, memory), server clients, server installation, domains, user accounts, groups, rights, directories, permissions, applications, printers, other OS, monitoring, maintenance, high speed switching, ATM, video, routers, fire walls. (YR).",
    "name": "High Speed Network Admin",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "435",
    "creditHours": 3,
    "description": "This course deals with the study of technologies used to design and implement multimedia web sites. Topics include web servers, HTML, CGI, scripting languages, Java applets, back-end database connectivity, web security, multimedia, XML. (F,W).",
    "name": "Web Technology",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "436",
    "creditHours": 3,
    "description": "This course introduces students to the development of software applications for programmable mobile and wireless intelligent hand-held devises. Topics covered include the different mobile development platforms, best practices in mobile user interaction design, software quality assurance in mobile environment, security and privacy issues, and context-aware computing.  Students will participate in a final project.",
    "name": "Mobile App Des & Impl",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "437",
    "creditHours": 3,
    "description": "Topics include an overview of the internet, congestion control, quality of service, internet multicasting, multimedia networking, mobile and wireless networks, vehicular networks, overlay networks, peer-to-peer networks, internet management (SNMP), and internet applications (web-HTTP and email-SMTP).",
    "name": "Advanced Networking",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "447",
    "creditHours": 3,
    "description": "This course will provide a broad-spectrum introduction to the fundamental principles of computer and network security. Topic will include security policies, models and mechanisms for confidentiality, integrity and availability, access control, authorization, cryptography and applications, threats and vulnerabilities in computer networks, key management, firewalls and security services in computer networks.",
    "name": "Intro Computr & Ntwrk Security",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "450",
    "creditHours": Array [
      3,
      4,
    ],
    "description": "Introduction to computer operating systems. Process control, threads, concurrency, memory management, virtual memory, uniprocessor, multiprocessor, and real-time scheduling, I/O management, disk scheduling, file management, distributed processing, client/server, clusters, distributed process management,security. (F,W).",
    "name": "Operating Systems",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "451",
    "creditHours": 3,
    "description": "Basic geometrical concepts: graphics output primatives, two-dimensional transformations, windowing and clipping, three-dimensional viewing, visible surface detection methods, and graphical user interfaces. (F).",
    "name": "Computer Graphics",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "452",
    "creditHours": 3,
    "description": "This course introduces basic techniques for digital animation, computer and video games, and web multimedia. Topics include the process of creating animated video clips from start to finish, including story creation, storyboarding, modeling, animation, and post-production; several key techniques for video editing and motion generation, including keyframe, motion capture editing, collision detection, particle systems, physical simulation, and real-time rendering; techniques for web animation and multimedia; and internet gaming.",
    "name": "Inf Vis & Multimedia Gaming",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "456",
    "creditHours": 3,
    "description": "This course covers the core tenets of the Microsoft Foundation Class (MFC) or similar package and Windows programming. The emphasis will be on the relationship between Windows Operating System and MFC. Windows OS has three major components: user, graphics device interface (GDI), and kernel. User is a module that controls input devices, GDI is a module that services output devices, and kernel controls internal resources. These three components are called the API and communicate with MFC. Projects will be assigned to simulate the major components of API using MFC. (YR)",
    "name": "Windows Programming",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "467",
    "creditHours": 4,
    "description": "This course is a continuation of Digital Forensics I and will focus on Internet Forensics. Students will examine in-depth concepts in Internet evidence collection and preservation, as well as applications of contemporary commercial forensic investigative software.",
    "name": "Digital Forensics II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "474",
    "creditHours": 3,
    "description": "Principles of language compilation.  Introduction to formal languages.  Lexical analysis, top-down and bottom-up parsing, code generation and optimization.  Error handling and symbol table management.  Run-time storage management. Programming language design.  Introduction to compiler-writing tools such as LEX and YACC. (F,W).",
    "name": "Compiler Design",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "475",
    "creditHours": 3,
    "description": "The focus of this course is on management issues related to modern software engineering practice. Students read and discuss papers written by master software engineering professionals. Seminar topics discussed include: management of software engineering processes, software measurement, software engineering ethics, and legal issues related to professional practice. (W, S).",
    "name": "Software Engineering Seminar",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "476",
    "creditHours": 3,
    "description": "This course focuses on design patterns in object-oriented programming. This course begins with an overview of UML and a review of object-oriented programming and then moves on to various structural, behavioral and creational patterns, including: facades, adaptors, bridges, factories and the template method. Analysis of case studies will also be discussed. Using various modern software tools, students will apply various design patterns to real-world software design problems to gain complete practical understanding. (F,W)",
    "name": "Soft Arch & Design Patterns",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "479",
    "creditHours": 3,
    "description": "This course introduces students to basic concepts and methods of artificial intelligence from a computer science perspective. Emphasis of the course will be on the selection of data representations and algorithms useful in the design and implementation of intelligent systems. The course will contain an overview of one AI language and some discussion of important applications of artificial intelligence methodology. (S).",
    "name": "Intro to Artificial Intel",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "481",
    "creditHours": 3,
    "description": "This course covers basic computational aspects of learning to perform a task and improve with experience. Topics include learning frameworks and problem formulations; standard models, methods, computational tools, algorithms and modern techniques; and methodologies to evaluate learning ability and to automatically select optimal models. The main focus is on computer science (e.g., basic runtime, space and complexity analysis, programming, and empirical evaluations?. Simple applications to areas such as computer vision, natural language processing (NLP), and robotics will also motivate the course material. (W)",
    "name": "Computational Learning",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4851",
    "creditHours": 3,
    "description": "This course covers basics of data security and privacy techniques, which can facilitate the use of data in a secure and privacy-sensitive way. Topics include security and privacy challenges due to data collection and analytics, technologies and strategies for data security and privacy (access control mechanism, integrity policy, cryptography and encryption, notice and consent, anonymization or de-identification, deletion and non-retention). (W)",
    "name": "Data Security and Privacy",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "487",
    "creditHours": 3,
    "description": "This course deals with the study of the technology, science and art in the creation of computer games. The focus of the course will be hands-on development of computer games. Students will study a variety of software technologies relevant to computer game design, including programming languages, scripting languages, operating systems, files systems, networks, simulation engines and multi-media design systems. Lecture topics will be taken from several areas of computer science: simulation and modeling, computer graphics, artificial intelligence, real-time processing, game theory, software engineering, human computer interaction, graphic design and game aesthetics. (F).",
    "name": "Computer Game Design & Implem",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "488",
    "creditHours": 3,
    "description": "This course is a continuation of the material studied in CIS 487. The focus of the course will be hands-on development of computer game development tools (e.g. game engines). Students will study a variety of software technologies relevant to computer game design, including: 3D graphics, computer animation, data-driven game design, multiplayer game programming, and game AI. Lecture topics will be taken from several areas of computer science: simulation and modeling, computer graphics, artificial intelligence, game theory, software engineering, human computer interaction and game content development. (W)",
    "name": "Computer Game Design II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "490",
    "creditHours": Array [
      1,
      3,
    ],
    "description": "This course is intended for seniors and graduate-level students in CIS.  For specific topic, consult current semester's Schedule of Classes. (OC).",
    "name": "Advanced Topics",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "490G",
    "creditHours": 3,
    "description": "Topic Title: Special Topics: Wireless and Mobile Computing Security: The course focuses on security and privacy issues in the area of wireless networks and mobile computing such as cellular networks, wirelss LANs, connected vehicles, smart and mobile devices, sensors and sensor networks, loT, etc. The course will first present an overview of wireless communication and wireless systems, then foucs on attacks, discuss proposed solutions and their limitations. Topics of this course include: (1) introduction to security primitives wirelessand wireless networks; (2) security issues in single-hop wireless networks that include cellular networks, RFID, modern vehicle, smartphone security; (3) security issues in multi-hop wirless network that include Mobile As Hoc network, wireless sensor network and vehicular network security.(YR)",
    "name": "Adv Topics in Computer Science",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "491",
    "creditHours": Array [
      1,
      4,
    ],
    "description": "Provides the advanced student with the opportunity to undertake a research project under the supervision of a faculty member.  At least two weeks prior to registration in the semester when such a course is to be elected, an interested student must submit to the CIS chair and one CIS faculty member a written request for permission to elect a research course on the appropriate form available in the CIS Office.  The request will include a description of the proposed research project.  The CIS chair will review the proposal with faculty members to ascertain availability of relevant faculty supervision and to establish appropriate credit.  Grades will be granted on a Pass/Fail (S/E) basis exclusively. (F,W,S).",
    "name": "Research Project I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "492",
    "creditHours": Array [
      1,
      4,
    ],
    "description": "This course is a second registration for a research project in CIS. (F,W,S).",
    "name": "Research Project II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "493",
    "creditHours": Array [
      1,
      4,
    ],
    "description": "Readings or analytical assignments in accordance with the needs and interests of those enrolled and agreed upon by the student and an instructor, which shall not duplicate a formal course offering. Permission of instructor required. (F,W,S).",
    "name": "Independent Study I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "494",
    "creditHours": Array [
      1,
      4,
    ],
    "description": "This course is a second registration for an independent study in CIS. Permission of instructor required. (F,W,S).",
    "name": "Independent Study II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "495",
    "creditHours": 4,
    "description": "Students participate in the design and implementation of a major software project. Seminar topics discussed include: computing ethics and professional practice in computer science. (F,W,S).",
    "name": "Design Seminar",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4951",
    "creditHours": 2,
    "description": "Students participate in the design and implementation of a major software project. Seminar topics discussed include: computing ethics and professional practice. (F,W,S)",
    "name": "Design Seminar I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4952",
    "creditHours": 2,
    "description": "Students continue to participate in the design and implementation of a major software project. Seminar topics discussed include: computing ethics and professional practice. (F,W,S)",
    "name": "Design Seminar II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "496",
    "creditHours": 4,
    "description": "Software engineering students participate in the design and implementation of a major software project.  Seminar topics discussed include: computing ethics and professional practice in software engineering.",
    "name": "Design Seminar for SE",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4961",
    "creditHours": 2,
    "description": "Software engineering students participate in the design and implementation of a major software project. Seminar topics discussed include: computing ethics and professional practice in software engineering. (F,W,S)",
    "name": "Design Seminar for SE I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4962",
    "creditHours": 2,
    "description": "Software engineering students continue to participate in the design and implementation of a major software project. Seminar topics discussed include: computing ethics and professional practice in software engineering.",
    "name": "Design Seminar for SE II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4971",
    "creditHours": 2,
    "description": "Data science students participate in the design and implementation of a major data science project. Seminar topics discussed include: computing ethics and professional practice in data science. (F, W, S)",
    "name": "Cap Sem for Data Sci I",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "4972",
    "creditHours": 2,
    "description": "Data science students continue to participate in the design and implementation of a major data science project. Seminar topics discussed include: computing ethics and professional practice in data science. (F, W, S)",
    "name": "Cap Proj for Data Sci II",
    "subjectCode": "CIS",
  },
  Object {
    "courseNumber": "499",
    "creditHours": 1,
    "description": "Student works with industrial sponsor in area of CIS. Approval of Internship Coordinator required. (F,W,S).",
    "name": "Internship",
    "subjectCode": "CIS",
  },
]
`;
